PyPortal_AlarmClock/code.py

```diff
@@ -0,0 +1,487 @@
+        pass
+
+
+    def exit(self):
+        """Just before the state exits."""
+        clear_splash()
+
+
+class Time_State(State):
+    """This state manages the primary time display screen/mode"""
+
+    def __init__(self):
+        super().__init__()
+        self.background_day = 'main_background_day.bmp'
+        self.background_night = 'main_background_night.bmp'
+        self.refresh_time = None
+        self.update_time = None
+        self.weather_refresh = None
+        text_area_configs = [dict(x=88, y=30, size=5, color=0xFFFFFF, font=large_font),
+                             dict(x=210, y=10, size=5, color=0xFF0000, font=medium_font),
+                             dict(x=88, y=65, size=6, color=0xFFFFFF, font=small_font)]
+        self.text_areas = create_text_areas(text_area_configs)
+        self.weather_icon = displayio.Group(max_size=1)
+        self.weather_icon.x = 88
+        self.weather_icon.y = 20
+        self.icon_file = None
+
+        self.snooze_icon = displayio.Group(max_size=1)
+        self.snooze_icon.x = 270
+        self.snooze_icon.y = 58
+        self.snooze_file = None
+
+        # each button has it's edges as well as the state to transition to when touched
+        self.buttons = [dict(left=320, top=50, right=240, bottom=120, next_state='settings'),
+                        dict(left=320, top=155, right=240, bottom=220, next_state='mugsy')]
+
+
+    @property
+    def name(self):
+        return 'time'
+
+
+    def adjust_backlight_based_on_light(self, force=False):
+        """Check light level. Adjust the backlight and background image if it's dark."""
+        global low_light
+        if light.value <= 1000 and (force or not low_light):
+            pyportal.set_backlight(0.01)
+            pyportal.set_background(self.background_night)
+            low_light = True
+        elif force or (light.value >= 2000 and low_light):
+            pyportal.set_backlight(1.00)
+            pyportal.set_background(self.background_day)
+            low_light = False
+
+
+    def tick(self, now):
+        global alarm_armed, snooze_time, update_time
+
+        # is the snooze button pushed? Cancel the snooze if so.
+        if not snooze_button.value:
+            if snooze_time:
+                self.snooze_icon.pop()
+            snooze_time = None
+            alarm_armed = False
+
+        # is snooze active and the snooze time has passed? Transition to alram is so.
+        if snooze_time and ((now - snooze_time) >= snooze_interval):
+            change_to_state('alarm')
+            return
+
+        # check light level and adjust background & backlight
+        self.adjust_backlight_based_on_light()
+
+        # only query the online time once per hour (and on first run)
+        if (not self.refresh_time) or ((now - self.refresh_time) > 3600):
+            try:
+                pyportal.get_local_time(location=secrets['time_location'])
+                self.refresh_time = now
```

Comment:
could this use "timezone" instead of "time_location" I think timezone is more commonly used.